### ISSUE TITLE
I trained the model, how this happened?

### ISSUE THREAD

**Issue by luohao123** [duwt/OrienMask#12]
![Screenshot from 2021-09-01 12-08-59](https://user-images.githubusercontent.com/49749220/131610369-dbc5e87c-20f9-4d8d-a2d4-531bb0f17ec5.png)

As you can see, some mask is right, but some mask is shifted, do you know why?

**Comment by luohao123** [comment]
I have trained about 60000 iterations, and the loss about 60, any idea?

**Comment by duwt** [comment]
(1) I'm not sure what is happening about the shifted masks. Do you edit some codes of post-processing or visualization?
(2) The loss is normal.

**Comment by luohao123** [comment]
@duwt  thank u for your reply.

I modified your postprocess make it much more simpler. And merged it with loss since there calculation are almost same.

the rest of thing is identical, and now some mask normal, some abnormal, this is really hard to debug.

Here is my simplified version postprocess:

```
 # recover boxes based on anchors
            print('scale wh', self.scale_wh)
            pred_wh = pred_wh.detach().exp()
            pred_boxes[..., 0:2] = (
                pred_xy.detach() + self.grid_mesh_xy).view(nB, -1, 2)
            pred_boxes[..., 2:4] = (
                pred_wh * self.grid_anchors.view(1, -1, 1, 1, 2)).view(nB, -1, 2)

            # for orien mask recover
            orien_xy = pred_boxes[..., 0:2].clone()
            orien_wh = (pred_wh * self.normalized_anchors.view(1, -1, 1, 1, 2)).view(nB, -1, 2) * self.grid_wh

            # recover boxes
            pred_boxes[..., 0:2] = pred_boxes[..., 0:2] * self.scale_wh
            pred_boxes[..., 2:4] = pred_boxes[..., 2:4] * self.scale_wh
            pred_obj = pred_obj.view(nB, -1, 1)
            pred_cls = pred_cls.view(nB, -1, self.num_classes)
            pred_boxes_obj_cls = torch.cat(
                [pred_boxes, pred_obj, pred_cls], dim=-1)

            # -> nb, na, 2, nh, nw
            pred_orien = pred_orien.permute(0, 1, 4, 2, 3)
            pixel_orien = pred_orien * self.grid_anchors.view(-1, 2, 1, 1) / 2

            pixel_orien += self.base_xy
            # print(pixel_orien.shape)
            return pred_boxes_obj_cls, pixel_orien, orien_xy, orien_wh, self.dets_anchor_idx
```

and here is how I recover boxes and masks:

```
def postprocess_mask(prediction, preds_oriens, xys, whs, dets_anchor_idxes, num_classes, conf_thre=0.7, nms_thre=0.45, orien_thre = 0.3):
    box_corner = prediction.new(prediction.shape)
    box_corner[:, :, 0] = prediction[:, :, 0] - prediction[:, :, 2] / 2
    box_corner[:, :, 1] = prediction[:, :, 1] - prediction[:, :, 3] / 2
    box_corner[:, :, 2] = prediction[:, :, 0] + prediction[:, :, 2] / 2
    box_corner[:, :, 3] = prediction[:, :, 1] + prediction[:, :, 3] / 2
    prediction[:, :, :4] = box_corner[:, :, :4]

    output = [None for _ in range(len(prediction))]
    pred_masks = [None for _ in range(len(prediction))]
    for i, image_pred in enumerate(prediction):
        xy = xys[i]
        wh = whs[i]
        dets_anchor_idx = dets_anchor_idxes[i]
        pred_orien = preds_oriens[i]

        print('xy', xy.shape)
        print('pred_orien', pred_orien.shape)

        # If none are remaining => process next image
        if not image_pred.size(0):
            continue
        # Get score and class with highest confidence
        class_conf, class_pred = torch.max(
            image_pred[:, 5: 5 + num_classes], 1, keepdim=True)

        # conf_mask = (image_pred[:, 4] *
        #              class_conf.squeeze() >= conf_thre).squeeze()
        conf_mask = (image_pred[:, 4] >= conf_thre).squeeze()

        # _, conf_mask = torch.topk((image_pred[:, 4] * class_conf.squeeze()), 1000)
        # Detections ordered as (x1, y1, x2, y2, obj_conf, class_conf, class_pred)
        detections = torch.cat(
            (image_pred[:, :5], class_conf, class_pred.float()), 1)
        detections = detections[conf_mask]

        xy = xy[conf_mask]
        wh = wh[conf_mask]
        dets_anchor_idx = dets_anchor_idx[conf_mask]

        if not detections.size(0):
            continue

        nms_out_index = torchvision.ops.batched_nms(
            detections[:, :4],
            detections[:, 4] * detections[:, 5],
            detections[:, 6],
            nms_thre,
        )
        detections = detections[nms_out_index]

        xy = xy[nms_out_index]
        wh = wh[nms_out_index]
        anchor_idx = dets_anchor_idx[nms_out_index]
        print('xy', xy.shape)
        print('wh', wh.shape)
        print('dets_anchor_idx', anchor_idx)

        print(pred_orien[anchor_idx, 0].shape)
        print('xy[:, 0].view(-1, 1, 1)', xy[:, 0].view(-1, 1, 1))
        print('xy[:, 0].view(-1, 1, 1)', wh[:, 1].view(-1, 1, 1))
        masks = ((torch.abs(pred_orien[anchor_idx, 0] - xy[:, 0].view(-1, 1, 1)) <
                  orien_thre * wh[:, 0].view(-1, 1, 1)) &
                 (torch.abs(pred_orien[anchor_idx, 1] -  xy[:, 1].view(-1, 1, 1)) <
                  orien_thre * wh[:, 1].view(-1, 1, 1)))
        print(masks.shape)
        # for j, m in enumerate(masks):
        #     a = m.cpu().numpy().astype(np.uint8) * 64
        #     cv2.imshow(f'aa{j}', a)
            # cv2.waitKey(0)

        if output[i] is None:
            output[i] = detections
            pred_masks[i] = masks
        else:
            output[i] = torch.cat((output[i], detections))
            pred_masks[i] = torch.cat((pred_masks[i], pred_masks))
    return output, pred_masks
```

Now, some mask normal some not, do u see any logic error in my code? (I think the boxes part is normal since boxes are all right)

**Comment by duwt** [comment]
It seems like that some large object masks are 1/4 height and width of the image but their shapes are normal. I can not imagine what kind of bug produces such errors.

**Comment by luohao123** [comment]
@duwt  the factors effected mask I can saw only these:

```
masks = ((torch.abs(pred_orien[anchor_idx, 0] - x_centers) <
                  self.orien_thresh * det_width * self.grid_sizes[anchor_idx, 0].view(-1, 1, 1)) &
                 (torch.abs(pred_orien[anchor_idx, 1] - y_centers) <
                  self.orien_thresh * det_height * self.grid_sizes[anchor_idx, 1].view(-1, 1, 1)))
```

which are `pred_orien`, `xy_centers`, `wh`, `anchor_idx` and `grid_size`,

since some mask is normal, so `pred_orien` definitely should be right. in my calculation, the grid_size already did when calculate `xy_centers` and `wh`, so grid_size should be right,

the only potential problem is in `xy_centers` and `wh` or `anchor_idx`.

Here is how I calculate `xy_centers`:

```python
pred_wh = pred_wh.detach().exp()
pred_boxes[..., 0:2] = (
    pred_xy.detach() + self.grid_mesh_xy).view(nB, -1, 2)
pred_boxes[..., 2:4] = (
    pred_wh * self.grid_anchors.view(1, -1, 1, 1, 2)).view(nB, -1, 2)

# for orien mask recover
orien_xy = pred_boxes[..., 0:2].clone()
orien_wh = (pred_wh * self.normalized_anchors.view(1, -1, 1, 1, 2)).view(nB, -1, 2) * self.grid_wh
```

Does orien_xy and orien_wh looks normal? The only last problem I came to mind now is anchor error, but I can not know which anchor got error since boxes looks all right.

I calculate one level anchor_idx by:

```python
anchor_idx = torch.tensor(self.anchor_mask, device=self.device).view(
    self.num_anchors, 1, 1)
anchor_idx = anchor_idx.expand(b, self.num_anchors, h, w).contiguous()
self.dets_anchor_idx = anchor_idx.view(b, -1)
```

this seems no problem?

this is identical with your code:

```python
for m, nA, nH, nW in zip(anchor_mask, self.num_anchors, self.nHs, self.nWs):
            # base coordinates of orientations (unit size = grid)
            base_y, base_x = torch.meshgrid(
                [torch.arange(self.image_h, device=self.device, dtype=torch.float) / self.image_h * nH,
                 torch.arange(self.image_w, device=self.device, dtype=torch.float) / self.image_w * nW])
            self.base_xy[m] = torch.stack([base_x, base_y], dim=0)

            # base coordinates of grid centers (unit size = grid)
            grid_y, grid_x = torch.meshgrid(
                [torch.arange(nH, device=self.device, dtype=torch.float),
                 torch.arange(nW, device=self.device, dtype=torch.float)])
            self.grid_y.append(grid_y.expand(nA, nH, nW).contiguous())
            self.grid_x.append(grid_x.expand(nA, nH, nW).contiguous())

            # anchor indices of all predictions
            anchor_idx = torch.tensor(m, device=self.device).view(nA, 1, 1)
            self.anchor_idx.append(anchor_idx.expand(nA, nH, nW).contiguous())
 self.dets_anchor_idx = torch.cat([anchor_idx.view(-1) for anchor_idx in self.anchor_idx], dim=0)
```

**Comment by duwt** [comment]
I suggest that taking an image as input to see what is the difference between the intermediate variables of our codes and your version. It should be more efficient to debug.

**Comment by luohao123** [comment]
@duwt yeah, but can you notice any mistake in my simplified version? cause it looks like identical, and I just made input to rectangle (dynamic grid sizes based on input size)

**Comment by duwt** [comment]
I have no idea. Furthermore, I think the error is not caused by this part since some masks are normal. If all anchors are processed in the same way, the masks will be all correctly or mistakenly predicted.

**Comment by luohao123** [comment]
@duwt  thank u ! I figured it out finally. And now I reproduced result with more simplified code and rectangle inputs:

![image](https://user-images.githubusercontent.com/49749220/132081700-edad2188-acdd-43ad-bb45-26c892b27345.png)


not fully trained though, am using same backbone as YOLOX, very fast.

**Comment by luohao123** [comment]
@duwt  BTW, did you tried using ciou as box loss rather than vanilla yolov3 xywh loss? I found the mask actually works well in this orien way, so that, if boxes and detection can be more accurate, that will make this detector more powerful. One simple is to apply this loss and add some iou aware branch.

**Comment by duwt** [comment]
@luohao123 Nice! We have not attempted more improvements for YOLOv3 and the mask quality should be better if detection results are improved.

**Comment by luohao123** [comment]
@duwt I tried add ciou now, do u think this loss is normal?

![image](https://user-images.githubusercontent.com/49749220/132082665-3f4fa52a-25e2-41be-afd4-fc8d24eab12b.png)


Do u have any further ideas to strengthen orien? Would you like share some thoughts?

**Comment by duwt** [comment]
@luohao123 We have grafted orientation maps to the anchor-free detector FCOS. It is more general but the accuracy is only on par with the current result. We are trying to optimize the framework but the priority is not that high. We will update and release codes if any progress has been made.

**Comment by luohao123** [comment]
@duwt  focus is a good choice,  how about yolox?

**Comment by duwt** [comment]
@luohao123 We have no plan for YOLOX yet.

**Comment by luohao123** [comment]
@duwt  how's the performance compare with centermask when using fcos as base detector?

**Comment by duwt** [comment]
@luohao123 CenterMask-Lite with ResNet-50 backbone obtains 32.6 mask AP (val) at a speed of 27.4 fps in our platform. Currently, OrienMask-FCOS has 34.5 at a speed of around 37.8 fps.

**Comment by luohao123** [comment]
@duwt  Indeed, it seems promising. I found that if I change a backbone of orienmask, the mask performance seems not good, do u know what might effect this?

![Screenshot from 2021-09-06 13-00-12](https://user-images.githubusercontent.com/49749220/132163038-0b896e12-fc44-424d-b6bd-0cae84984250.png)

**Comment by duwt** [comment]
@luohao123 I'm not sure if there are any other modifications except for the backbone, like the training data, training scheduler, anchor settings, and detector head. Generally speaking, it should be sufficiently trained due to the compact mask representation.

**Comment by luohao123** [comment]
@duwt data and augmentation same, except without color jitter. And training scheduler am using sgd + warmup + cosine decay scheduler.

the anchors are same, another modification is am using rectangle input which up to 412-800 as maxium. the box detection seems normal.

Another problem I found is that same object might have multiple boxes which even can not be filtered out with nms_thresh=0.6, don't know why.

Do u have a plan release fcos version? it would be better if you build upon detectron2 so that can be compared with many popular model under same augmentation && input condition.

what's the orienmask output dimension BTW in anchor-free detector?

**Comment by duwt** [comment]
(1) Multiple boxes for the same object may be caused by inaccurate classification. If two boxes have different class ids but both logits are above the threshold, they will be kept in the results.
(2) I guess that the model is not fully trained although the boxes seem good (the AP metric may not be that appealing). The YOLOv3 needs to be trained over 270 epochs in COCO so we initialize the backbone part with a pre-trained detector to speed up convergence. If the backbone is replaced with an ImageNet pre-trained classifier, it should be equipped with a longer scheduler.
(3) Our FCOS model uses almost the same backbone and scheduling as the real-time BlendMask ([link](https://github.com/aim-uofa/AdelaiDet/blob/master/configs/BlendMask/550_R_50_3x.yaml)). The implementation is based on our current framework but I have no clear plan about when the codes will be released. (Now I am working on another project for graduation.)

**Comment by luohao123** [comment]
@duwt Hoping for your updates, please keep dev on this method, it's very promising. I'll update my exp result if i have any progress

**Comment by duwt** [comment]
@luohao123 Thanks for your interest in our work. I hope more improvements could be made and then share the progress.

**Comment by luohao123** [comment]
@duwt New update:

I found that pretrained model is the key to train, I using YOLOX backbone as init weights, and now after 4000 iters I can have a good detection and full mask without too many holes.

Also, I eliminated xy wh loss, replace with ciou, faster to converge.

**Comment by luohao123** [comment]
![Screenshot from 2021-09-07 18-27-28](https://user-images.githubusercontent.com/49749220/132329791-896a70ee-e330-4017-a8b2-d2b08a8494fe.png)

9000 iters can have a good mask

**Comment by luohao123** [comment]
@duwt  I found a interesting thing:

![Screenshot from 2021-09-08 14-55-44](https://user-images.githubusercontent.com/49749220/132461085-4223ce3f-2005-413a-8ed1-54c06f7f2f84.png)

I found that in small objects, the mask will floating, although their box are fine, Do u know what caused this problem?

**Comment by duwt** [comment]
@luohao123 If all masks float, check the padding and cropping operations for the input image and predicted masks. If only small masks happen, make sure the post-processing is consistent for all anchors.

**Comment by luohao123** [comment]
@duwt  Not all small mask floating.

this pic is fine:

![Screenshot from 2021-09-08 15-41-16](https://user-images.githubusercontent.com/49749220/132467353-244d2161-e929-430d-b528-f1a194f76307.png)
![Screenshot from 2021-09-08 15-40-57](https://user-images.githubusercontent.com/49749220/132467367-50c119af-ad54-418c-9ae1-c26630653661.png)

do u know, if the box is not accurate, does box will effect mask quality?

**Comment by luohao123** [comment]
@duwt  BTW, do u know how to adopt these anchors in orienmask map from anchor-based model to anchor-free model such as fcos?

**Comment by duwt** [comment]
(1) I'm not sure the reason for floating masks. Mask quality is not sensitive to box predictions but they are learned within a united network.
(2) In anchor-free settings, we group objects by their sizes and assign them to different orientation maps. The mechanism is similar to the detector.

**Comment by luohao123** [comment]
@duwt In yolov3 we have 9 anchors, so that output dimension could be 9, 544, 544, 2

what's the output dimension in anchorfree based model?

**Comment by duwt** [comment]
@luohao123 The output dimension for FCOS is (N, H, W, 2) where N is the number of orientation maps. It depends on your settings of "sizes_of_interest_orien", which is similar with "sizes_of_interest" in FCOS. In our experiments, N should be at least 9 to obtain acceptable results.

**Comment by luohao123** [comment]
@duwt is it different base sizes of orienmaps? in yolov3 we have 3 sizes and 3 ratios, but in fcos, is there need 9 scales to predict all sizes targets? Then how to assign them respectively same bbox assignment?

**Comment by duwt** [comment]
@luohao123 The assignment strategy is similar to FCOS, which is based on the width and height of bounding boxes. If you are interested in this extension, feel free to explore and we have merely built a plain baseline on FCOS.

**Comment by luohao123** [comment]
@duwt if you want assistant on MR or doing experiement, I'd like to help, to help you migrate to more anchorfree and bring up baseline togather, it would be great if you invite me to your private baseline on fcos.

I actually now have another version on OrienMask and quite skilled how to do with Orien postprocessing, it would be helpful to make further experiment togather, I'd obey your protocals to MR and keep private until you release officially.

**Comment by duwt** [comment]
@luohao123 Thanks for your generous help. The problem is that our experimental codes should be organized carefully to release but I have no sufficient time to do. Since the anchor-free setting makes the orientation maps less coherent with the detector, the current result on FCOS is not that appealing and should be explored further. If you believe it is promising, feel free to try.

**Comment by luohao123** [comment]
@duwt I'll wait your release then, I trying to optimize this anchor-based way at the moment.

**Comment by luohao123** [comment]
![Screenshot from 2021-09-10 12-38-52](https://user-images.githubusercontent.com/49749220/132800566-a7e8c5df-07f8-4478-95e3-6b61e6303d34.png)

@duwt I fixed the mask floating issue now.

**Comment by luohao123** [comment]
@duwt Did u found a way to improve the mask quality,  I notice that the model inference on some strangeness image, the mask would be out of box and seems not very tight.

**Comment by pupu-chenyanyan** [comment]
@luohao123 Hi, Have you ever had this problem of  "pred_wh not finite"? I guess this maybe caused by WH loss. And I noticed that you  eliminated xy wh loss, replace with ciou. Can you give me some guidance about  removing wh loss?Thank you very much !

**Comment by luohao123** [comment]
@pupu-chenyanyan sorry, I didn't track this project for a long time, maybe you can checkout someone else who add orient yolo, search yolov7 on github.

**Comment by theyunfeng** [comment]
> ![Screenshot from 2021-09-10 12-38-52](https://user-images.githubusercontent.com/49749220/132800566-a7e8c5df-07f8-4478-95e3-6b61e6303d34.png)
> 
> @duwt I fixed the mask floating issue now.

May I ask how you solved this problem? If it's convenient, could you share the GitHub code?
